src/sagemaker_xgboost_container/algorithm_mode/train.py

```diff
@@ -203,14 +201,100 @@ def train_job(train_cfg, train_dmatrix, val_dmatrix, train_val_dmatrix, model_di
     else:
         train_cfg.pop('eval_metric', None)
 
-    # Set callback evals
-    watchlist = [(train_dmatrix, 'train')]
-    if val_dmatrix is not None:
-        watchlist.append((val_dmatrix, 'validation'))
+    logging.info("Train matrix has {} rows and {} columns".format(train_dmatrix.num_row(), train_dmatrix.num_col()))
+    if val_dmatrix:
+        logging.info("Validation matrix has {} rows".format(val_dmatrix.num_row()))
+
+    try:
+        kfold = train_cfg.pop("_kfold", None)
+
+        if kfold is None:
+            xgb_model, iteration, callbacks, watchlist = get_callbacks_watchlist(
+                train_cfg, train_dmatrix, val_dmatrix, model_dir, checkpoint_dir)
+            add_debugging(callbacks=callbacks, hyperparameters=train_cfg, train_dmatrix=train_dmatrix,
+                          val_dmatrix=val_dmatrix)
+
+            bst = xgb.train(train_cfg, train_dmatrix, num_boost_round=num_round-iteration, evals=watchlist,
+                            feval=configured_feval, early_stopping_rounds=early_stopping_rounds,
+                            callbacks=callbacks, xgb_model=xgb_model, verbose_eval=False)
+
+        else:
+            logging.info("Run {}-fold cross validation with {} rows".format(kfold, train_val_dmatrix.num_row()))
+
+            bst = []
+            evals_results = []
+
+            num_cv_round = train_cfg.pop("_num_cv_round", 1)
+            for cv_round in range(num_cv_round):
+                if num_cv_round > 1:
+                    logging.info("Run round {} cross validation".format(cv_round+1))
+
+                train_val_size = train_val_dmatrix.num_row()
+                row_index = np.random.permutation(range(train_val_size))
+                fold_size = round(train_val_size / kfold)
+                for fold in range(kfold):
```

Comment:
Can we simplify by using sklearn function ([RepeatedKFold](https://scikit-learn.org/stable/modules/generated/sklearn.model_selection.RepeatedKFold.html)) for the indices?

Reply:
Good to know. Will update. Thanks.

Reply:
Updated with `RepeatedKFold`.